### PR TITLE
[bridge indexer] - indexer refactoring

### PR DESCRIPTION
**Pull request by patrickkuo**
## Description 

Describe the changes or additions included in this PR.

## Test plan 

How did you test the new or updated feature?

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:81]: iiuc, this should be called `start_block`, right?

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:100]: 1. likewise a batch_writer api will be more efficiently here
2. for max throughput, we should leverage channels to process messages in a pipeline, as opposed to blocking here

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:119]: this part is a bit error prone. It's not hard to have two unrelated tasks with the same prefix.

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:117]: where is the logic to filter out finished tasks?

File: crates/sui-bridge-indexer/src/indexer_builder.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/indexer_builder.rs:229]: nit: return early if empty to reduce indentation

File: crates/sui-bridge-indexer/src/indexer_builder.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/indexer_builder.rs:283]: rename this to `start_data_retrieval`? it does more than "create" data channel

File: crates/sui-bridge-indexer/src/indexer_builder.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/indexer_builder.rs:344]: one suggestion to keep this file lean: move all ingestion framework specific things to another file. Or simply only keep traits here and move all implementations to their corresponding files

Comment (longbowlu) [inline crates/sui-bridge-indexer/src/indexer_builder.rs:344]: `PerTaskInMemProgressStore` is perhaps more self descriptive

File: crates/sui-bridge-indexer/src/indexer_builder.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/indexer_builder.rs:238]: can we add a TODO here to explain the opportunity to do these two writes in one DB transaction, or, update save_progress less frequently? (for better backfilling performance)

in particular, it's not clear to me whether we should separate `IndexerProgressStore` and `Persistent` traits

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:45]: this should not be in `*sui*_bridge_indexer.rs` file, right? it's shared between eth and sui

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:103]: add a comment that these two columns won't be updated

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:138]: shouldn't this be the current timestamp?

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:150]: what is `now`?

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:213]: I didn't read this part, assuming this is just copied over

File: crates/sui-bridge-indexer/src/main.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/main.rs:100]: `zed` everywhere else

File: crates/sui-bridge-indexer/src/main.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/main.rs:144]: `zed`

**Review by longbowlu — COMMENTED** [review]
another comment in error handling: in a few places the task return Result<()>, but I'm not sure we handle the errors correctly or quietly drop them on the floor. It seems easier to just panic up in the stack when this happens

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs
Comment (patrickkuo) [reply crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:130]: send should only fail when channel is closed, in that case the receiver should have panicked already.

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs
Comment (patrickkuo) [reply crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:271]: we will need another review of all metrics after all the changes are in, I will defer the name changed until then.

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs
Comment (patrickkuo) [reply crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:318]: this is carried over from previous implementation, we only interested in unfinalized deposit event.

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (patrickkuo) [reply crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:117]: in the sql query few lines below `.filter(columns::checkpoint.lt(columns::target_checkpoint))`

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (patrickkuo) [reply crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:119]: will add `-` after the prefix for now, will need to change this to a better solution which will require changing the schema, will add a todo here.

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (patrickkuo) [reply crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:45]: will do another house keeping PR to move things around

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (patrickkuo) [reply crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:138]: timestamp is defaulted to database current time if not provided, this is defined in the schema

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (patrickkuo) [reply crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:150]: now is DB current time

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs
Comment (longbowlu) [reply crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:144]: the receiver could be dropped silently without panicking.

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs
Comment (longbowlu) [reply crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:332]: but what the code is doing is being interested in unfinalized claims. Do we want `if !self.finalized`?

File: crates/sui-bridge-indexer/src/indexer_builder.rs
Comment (longbowlu) [reply crates/sui-bridge-indexer/src/indexer_builder.rs]: I'm not sure target checkpoint is enough, range feels safe. But will leave it up to you. Feel free to resolve this one if you want to keep target checkpoint only

File: crates/sui-bridge-indexer/src/indexer_builder.rs
Comment (longbowlu) [reply crates/sui-bridge-indexer/src/indexer_builder.rs:176]: i see you changed line 173, so it's exclusive?

Comment (longbowlu) [reply crates/sui-bridge-indexer/src/indexer_builder.rs:176]: i see you changed line 173, so it's exclusive?

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (longbowlu) [reply crates/sui-bridge-indexer/src/sui_bridge_indexer.rs]: sg, plz add a TODO

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (longbowlu) [reply crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:118]: not sure i follow. This is the latest code i'm seeing:

```
        let cp: Vec<models::ProgressStore> = dsl::progress_store
            .filter(columns::task_name.like(format!("{prefix} - %")))
            .filter(columns::checkpoint.lt(columns::target_checkpoint))
            .order_by(columns::checkpoint.desc())
            .load(&mut conn)?;
```

Comment (longbowlu) [reply crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:118]: not sure i follow. This is the latest code i'm seeing:

```
        let cp: Vec<models::ProgressStore> = dsl::progress_store
            .filter(columns::task_name.like(format!("{prefix} - %")))
            .filter(columns::checkpoint.lt(columns::target_checkpoint))
            .order_by(columns::checkpoint.desc())
            .load(&mut conn)?;
```

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (longbowlu) [reply crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:45]: can we plz add a TODO here

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (longbowlu) [reply crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:139]: til

File: crates/sui-bridge-indexer/src/indexer_builder.rs
Comment (patrickkuo) [reply crates/sui-bridge-indexer/src/indexer_builder.rs:224]: added a todo here, will change all the metrics in another PR

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (patrickkuo) [reply crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:118]: this query returns all task with name with `prefix` and checkpoint less than target checkpoint, which are all the unfinished tasks

File: crates/sui-bridge-indexer/src/indexer_builder.rs
Comment (patrickkuo) [reply crates/sui-bridge-indexer/src/indexer_builder.rs:337]: will do the move in another PR to reduce conflict with Bridger's PR

File: crates/sui-bridge-indexer/src/sui_bridge_indexer.rs
Comment (longbowlu) [reply crates/sui-bridge-indexer/src/sui_bridge_indexer.rs:118]: yeah this is obvious, now i'm not sure why i asked this question. thanks for explaining!

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:120]: use a larger value, this means it will retry for 30s, which is a bit short when it's being rate limited

File: crates/sui-bridge-indexer/src/eth_bridge_indexer.rs
Comment (longbowlu) [inline crates/sui-bridge-indexer/src/eth_bridge_indexer.rs:129]: ditto

**Review by longbowlu — APPROVED** [review]

**Pull request merged**
